src/categorical_algebra/DataMigration.jl

```diff
@@ -0,0 +1,159 @@
+""" Data Migration functors
+"""
+module DataMigration
+
+export Σ, Functor
+
+using ...Theories
+using ...Theories: CatDesc, AttrDesc, ob, hom, dom, codom
+using ...CSetDataStructures
+using ..FinSets, ..CSets, ..Limits, ..FreeDiagrams
+using ...Graphs
+using ...Present
+
+import ...CategoricalAlgebra.FreeDiagrams: FreeDiagram
+
+""" Functor
+
+A functor ``F: \\mathcal{C} \\to \\mathcal{D}`` consists of a map of objects and a 
+map of homomorphisms from a domain category ``\\mathcal{C}``  to a codomain
+category ``\\mathcal{D}``.
+"""
+struct Functor{Schema, Ob, Hom}
+  FOb::Dict{Ob, Ob}
+  FHom::Dict{Hom, Hom}
+  dom::Presentation{Schema, Symbol}
+  codom::Presentation{Schema, Symbol}
+end
+
+""" Inv(F::Functor)
+
+For a functor ``F: \\mathcal{C} \to \\mathcal{D}``, 
+returns a function which when given an object (resp. morphism)
+of ``\\mathcal{D}``, returns an array of objects (resp. morphisms) 
+of ``\\mathcal{C}`` in the preimage.
+"""
+function Inv(F::Functor{Schema, Ob, Hom}) where {Schema, Ob, Hom}
+  y::Union{Ob, Hom} -> begin
+    y isa Ob ? D = F.FOb : D = F.FHom
+    [k for (k,v) in D if v == y]
+  end
+end
```

Comment:
maybe better to make a struct InvFunctor and then overload calling

Reply:
Yeah, then you could define that as two methods instead of a using `isa` that feels more natural with the way that a functor is defined. We think of a functor as any thing that has an action on objects and an action on morphisms. The only downside I see to using this interface is that we would need to wrap things that are already callables. https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects If we chose an interface like `fapply(F::AbstractFunctor, X::Ob)` and `fapply(F::AbstractFunctor, f::Hom)` then we could use `fapply(f::Function, x::FinSet)` where `f` is just a regular julia function. 

I think that since our objects and homs live in wrapper types anyway, this isn't that big of an issue and the benefit of `F(X::Ob)` and `F(f::Hom)` as the minimal syntax necessary to apply a functor is worth it.